syntax = "proto3";

package schemas;

message Array {
    repeated float elements = 1;  // One-dimensional array of integers
}

message Matrix {
    repeated Array rows = 1;  // Two-dimensional array of integers
}

message Tensor {
    repeated Matrix layers = 1;  // Three-dimensional array of integers
}

message BlockTensor {
    repeated Tensor blocks = 1;  // Four-dimensional array of integers
}

message MultiDimensionalArray {
    // Union type to represent any dimensional array
    oneof multi_array {
        Array array = 1;
        Matrix matrix = 2;
        Tensor tensor = 3;
        BlockTensor block_tensor = 4;
    }
}

message ModelConfig {
    string model_name = 1;
    int32 pp_rank = 2;
    int32 layer_idx_start = 3;
    int32 layer_idx_end = 4;
    string master_url = 5;
    int32 next_pp_rank = 6;
}

message ForwardRequest {
    string uuid = 1;
    MultiDimensionalArray hidden_states = 2;
}

message StatusResponse {
    string msg = 1;
    int32 status = 2;
}

message ForwardResponse {
    string msg = 1;
    int32 status = 2;
    MultiDimensionalArray output = 3;
    float cost_time = 4;
}

message HealthResponse {
    string msg = 1;
    int32 status = 2;
}

message InitModelFlagResponse {
    bool msg = 1;
    int32 status = 2;
}

message Empty {}


service RPCService {
    rpc InitModel (ModelConfig) returns (StatusResponse);
    rpc Forward (ForwardRequest) returns (ForwardResponse);
    rpc Health (Empty) returns (HealthResponse);
    rpc InitModelFlag (Empty) returns (InitModelFlagResponse);
}
